src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/JavaScriptEncoder.cs

```diff
@@ -205,4 +214,187 @@ private static unsafe bool TryWriteEncodedSingleCharacter(int unicodeScalar, cha
+                return true;
+            return !_allowedCharacters.IsUnicodeScalarAllowed(unicodeScalar);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public unsafe override int FindFirstCharacterToEncode(char* text, int textLength)
+        {
+            if (text == null)
+            {
+                throw new ArgumentNullException(nameof(text));
+            }
+            return _allowedCharacters.FindFirstCharacterToEncode(text, textLength);
+        }
+
+        // The worst case encoding is 6 output chars per input char: [input] U+FFFF -> [output] "\uFFFF"
+        // We don't need to worry about astral code points since they're represented as encoded
+        // surrogate pairs in the output.
+        public override int MaxOutputCharactersPerInputCharacter
+        {
+            get { return 12; } // "\uFFFF\uFFFF" is the longest encoded form 
+        }
+
+        static readonly char[] s_b = new char[] { '\\', 'b' };
+        static readonly char[] s_t = new char[] { '\\', 't' };
+        static readonly char[] s_n = new char[] { '\\', 'n' };
+        static readonly char[] s_f = new char[] { '\\', 'f' };
+        static readonly char[] s_r = new char[] { '\\', 'r' };
+        static readonly char[] s_back = new char[] { '\\', '\\' };
+        static readonly char[] s_doubleQuote = new char[] { '\\', '"' };
+        static readonly char[] s_singleQuote = new char[] { '\\', '\'' };
+
+        // Writes a scalar value as a JavaScript-escaped character (or sequence of characters).
+        // See ECMA-262, Sec. 7.8.4, and ECMA-404, Sec. 9
+        // http://www.ecma-international.org/ecma-262/5.1/#sec-7.8.4
+        // http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf
+        public unsafe override bool TryEncodeUnicodeScalar(int unicodeScalar, char* buffer, int bufferLength, out int numberOfCharactersWritten)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+            // ECMA-262 allows encoding U+000B as "\v", but ECMA-404 does not.
+            // Both ECMA-262 and ECMA-404 allow encoding U+002F SOLIDUS as "\/"
+            // (in ECMA-262 this character is a NonEscape character); however, we
+            // don't encode SOLIDUS by default unless the caller has provided an
+            // explicit bitmap which does not contain it. In this case we'll assume
+            // that the caller didn't want a SOLIDUS written to the output at all,
+            // so it should be written using "\u002F" encoding.
+            // HTML-specific characters (including apostrophe and quotes) will
+            // be written out as numeric entities for defense-in-depth.
+            // See UnicodeEncoderBase ctor comments for more info.
+
+            if (!WillEncode(unicodeScalar))
+            { return TryWriteScalarAsChar(unicodeScalar, buffer, bufferLength, out numberOfCharactersWritten); }
```

Comment:
Nit: Odd newlines.

Reply:
Noticed this myself but left it as is given other code in this file (like the `DefaultJavaScriptEncoder`). The same applies to other feedback (like access modifiers). I can fix in both, but I would prefer to do all the cleanup in a separate PR (including things like unused using directives).